### ISSUE TITLE
[Serve] Allow WebSocket endpoints for ray deployments using FastAPI

### ISSUE THREAD

**Issue by juangtato-ds** [ray-project/ray#25363]
### Description

According to ray documentation we can expose HTTP endpoint using FastAPI in ray deployments (https://docs.ray.io/en/latest/serve/http-servehandle.html#serve-fastapi-http). The current implemetation of Ray limits this endpoints to use only HTTP (https://discuss.ray.io/t/fastapi-and-websockets/3723).

FastAPI allows to provide WebSockets endpoints (https://fastapi.tiangolo.com/advanced/websockets/).
When trying to expose WebSockets endpoints using FastAPI in Ray deployments the aplication starts without any issue. But, an error is returned when the client tries to access the WebSocket endpoint.

Here is a simple implementation of WebSocket endpoint over Ray deployment using the documentation pages related above:

```python
@serve.deployment(
    name="websocket-test",
    version='v1',
    route_prefix="/test",
    ray_actor_options={
        'num_cpus': 0
    }
)
@serve.ingress(app)
class WebsocketServer:

    def __init__(self):
        self._uuid = str(uuid.uuid4())

    @app.get('/')
    async def get(self):
        return HTMLResponse(html)

    @app.websocket("/ws")
    async def websocket_endpoint(self, websocket: WebSocket):
        await websocket.accept()
        while True:
            data = await websocket.receive_text()
            await websocket.send_text(f"Message text was: {data}")
```

The following error log is prompted by Ray:
```bash
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/juan/miniconda3/envs/dala-37/lib/python3.7/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 199, in run_asgi
    result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
  File "/home/juan/miniconda3/envs/dala-37/lib/python3.7/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/home/juan/miniconda3/envs/dala-37/lib/python3.7/site-packages/ray/serve/http_proxy.py", line 269, in __call__
    assert scope["type"] == "http"
AssertionError
INFO:     connection open
INFO:     connection closed
```

### Use case

We want to provide a WebSocket and let the user request to perform a long operation. In our case, this operation is composed by serveral phases, and each phase generates a piece of information that must be provided/returned to the user.

We have a narrow SLA, so waiting for the whole process to end is not an option. We need to provide the each piece of information as fast as it is generated. Our initial intend was to orchestrate something like the following pseudo code:

```python
async def the_process(self, websocket: WebSocket):
    await websocket.accept()
    input_data = await websocket.receive()
    # generate ID, accept job
    await websocket.send({'id': 'xxx', 'more-stuff': 'stuff'})

    partial_information_00 = await first_actor.process.remote(question)
    await websocket.send(partial_information_00)

    # Next calls can be done in parallel
    object_refs = [
    	second_actor.process.remote(partial_information_00),
    	third_actor.process.remote(partial_information_00)
    	# or calling other Ray deployments or tasks
    ]
    
    # Do a ray.wait pattern
    while object_refs:
    	done, object_refs = ray.wait(object_refs)
    	# await websocket.send iteration over done

    # Close websocket
```

**Comment by vincentteyssier** [comment]
Is it released? It is marked as done on public roadmap but I don't see any PR reference. That would be a really helpful addition for ChatGPT/langchain streaming API

**Comment by NavinKumarMNK** [comment]
Did it get updated?
My use case:
distribute video feeds from different camera into the server. 
server should be able to compute the feed in distributed way
Is there any other way to use web socket and do model distribution across ray cluster

**Comment by akshay-anyscale** [comment]
this is supported in ray serve now - https://docs.ray.io/en/latest/serve/http-guide.html#websockets

**Comment by jdixosnd** [comment]
The documented snippet does not work. The server fails the assert statement, which requires request type as ''http"

> (HTTPProxyActor pid=38382)   File "/anaconda/envs/azureml_py38/lib/python3.8/site-packages/ray/serve/_private/http_proxy.py", line 279, in __call__
> (HTTPProxyActor pid=38382)     assert scope["type"] == "http"

**Comment by ereduled** [comment]
As mentioned above, the assertion makes it impossible to use websockets with Ray, despite the FastAPI back-end being capable of supporting it. Should we raise a new issue for this? I believe this issue should be reopened because it is not possible to use it websocket endpoints right now.
